File: 백준/Gold/4803. 트리/트리.cc
#include <iostream>
#include <vector>
#include <queue>
using namespace std;
const char nl = '\n';
int n, m;
int tc = 1;
void solve(){
  vector<vector<int>> adj(n+1);
  vector<int> p(n+1);
  int u, v;
  while(m--){
    cin >> u >> v;
    adj[u].push_back(v);
    adj[v].push_back(u);
  }
  int ans = 0;
  for(int i = 1; i <= n; i++){
    if(p[i]) continue;
    bool isTree = 1;
    queue<int> q;
    q.push(i);
    while(!q.empty()){
      auto cur = q.front(); q.pop();
      for(auto nxt : adj[cur]){
        if(p[cur] == nxt) continue;
        if(p[nxt]){
          isTree = 0;
          continue;
        }
        p[nxt] = cur;
        q.push(nxt);
      }
    }
    ans += isTree;
  }
  cout << "Case " << tc++ <<": ";
  if(ans == 0) cout << "No trees.\n";
  else if(ans == 1) cout << "There is one tree.\n";
  else cout << "A forest of " << ans << " trees.\n"; 
}
int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);
  while(1){
    cin >> n >> m;
    if(n == 0) break;
    solve();
  }
}

File: 백준/Gold/4803. 트리/트리.py
import io, os, sys
from collections import deque
from itertools import count

input = io.BytesIO(os.read(0, os.fstat(0).st_size)).readline

def main() -> None:
  for t in count(1):
    n, m = map(int, input().split())
    if n == 0:
      break
    solve(n, m, t)
    
def solve(N: int, M: int, testCase: int) -> None:
  # input
  G = [ [] for _ in range(N+1) ]
  P = [ 0 for _ in range(N+1) ]

  for _ in range(M):
    u, v = map(int, input().split())
    G[u].append(v)
    G[v].append(u)
  
  # solve
  treeCount: int = 0
  
  for i in range(1, N+1):
    if P[i] != 0:
      continue
    isTree: bool = True
    Q: deque = deque()
    Q.append(i)
    
    while Q:
      cur: int = Q.pop()
      for nxt in G[cur]:
        if P[cur] == nxt:
          continue
        if P[nxt] != 0:
          isTree = False
          continue
        P[nxt] = cur
        Q.append(nxt)
    
    if isTree:
      treeCount += 1
  
  # output
  print(f"Case {testCase}: ", end="")
  if treeCount == 0:
    print("No trees.")
  elif treeCount == 1:
    print("There is one tree.")
  else:
    print(f"A forest of {treeCount} trees.")

main()